File: prac/bj_02468_safe_zone.py
"""
https://www.acmicpc.net/problem/2468
"""
import sys


def sol(m, min_h, max_h) -> int:
    n, level = len(m), min_h - 1
    safe_zones, visited = [], [[-1 for _ in range(n)] for _ in range(n)]

    def count_safe_zone() -> int:
        count = 0
        for j in range(n):
            for i in range(n):
                count += dfs(i, j)
        return count
    
    def dfs(x: int, y: int) -> int:
        to_visit = [(x, y)]
        area = 0
        while to_visit:
            i, j = to_visit.pop()
            if (i >= 0 and j >= 0 and i < n and j < n
                and visited[j][i] != level
                and m[j][i] > level):
                visited[j][i] = level
                area += 1
                to_visit += [
                    (i - 1, j),
                    (i + 1, j),
                    (i, j - 1),
                    (i, j + 1),
                    ]
        if area > 0:
            return 1
        else:
            return 0

    while level <= max_h:
        safe_zones.append(count_safe_zone())
        level += 1

    return max(safe_zones)


if __name__ == "__main__":
    readline = sys.stdin.readline
    n, m = int(readline()), []
    min_h, max_h = 100, 1

    for _ in range(n):
        row = list(map(int, readline().split()))
        min_h, max_h = min(min_h, min(row)), max(max_h, max(row))
        m.append(row)

    print(sol(m, min_h, max_h))

File: prac/bj_02468_safe_zone.kt
/* https://www.acmicpc.net/problem/2468 */

fun main() {
    println(SafeZone(Array<List<Int>>(readLine()!!.toInt()) {
        readLine()!!.split(" ").map { it.toInt() }
    }).simul())
}

class SafeZone(matrix: Array<List<Int>>) {
    val m = matrix
    val n = matrix.size

    val minHeight: Int
    val maxHeight: Int

    init {
        var (min, max) = Pair(m[0][0], m[0][0])
        for (row in m) {
            for (h in row) {
                if (h < min) {
                    min = h
                } else if (h > max) {
                    max = h
                }
            }
        }
        minHeight = min
        maxHeight = max
    }

    var safeZones = arrayListOf<Int>(1)
    var visited = Array<Array<Int>>(n) { Array<Int>(n) { -1 } }

    fun simul(): Int {
        for (level in (minHeight - 1) .. maxHeight) {
            val zones = simul_level(level)
            safeZones.add(zones)
        }
        return safeZones.sorted().lastOrNull()!!
    }

    fun simul_level(level: Int): Int {
        var zones = 0
        for (j in 0..(n-1)) {
            for (i in 0..(n-1)) {
                zones += visit(i, j, level)
            }
        }
        return zones
    }

    fun visit(i: Int, j: Int, level: Int): Int {
        if (i < 0 || j < 0 || i >= n || j >= n ||
            visited[j][i] == level || m[j][i] <= level) {
            return 0
        } else {
            visited[j][i] = level
            visit(i - 1, j, level)
            visit(i + 1, j, level)
            visit(i, j - 1, level)
            visit(i, j + 1, level)
            return 1
        }
    }
}